File: src/year2024/day2/go/part2.go

package main


import (
    "fmt"
    "bufio"
    "strings"
    "strconv"
    "os"
)


func main() {
    file, _ := os.Open("../prod2.txt")
    defer file.Close()
    var ans int


    var lines []string
    scanner := bufio.NewScanner(file)
    for scanner.Scan() {
        lines = append(lines, scanner.Text())
    }

    for _, line := range lines {
        nums := parse_line(line)
        var safe = false
        if is_safe(nums) {
            safe = true
        } else {
                for i := 0; i < len(nums); i++ {
                    new_nums := make([]int, 0, len(nums)-1)
                    new_nums = append(new_nums, nums[:i]...)
                    new_nums = append(new_nums, nums[i+1:]...)
                    if is_safe(new_nums) {
                        safe = true
                        break
                    }
                }
        }

        if safe {
            ans++
        }
    }

    fmt.Println("Ans: ", ans)
}


func parse_line(line string) []int {
    nums := strings.Fields(line)
    parsed_nums := make([]int, len(nums))

    for i, c := range nums {
        if val, err := strconv.Atoi(c); err == nil {
            parsed_nums[i] = val
        }
    }
    return parsed_nums
}

func is_safe(nums []int) bool {
    allPositive := true
    allNegative := true
    maxThree := true
    minOne := true

    for i := 1; i < len(nums); i++ {
        diff := nums[i-1] - nums[i]

        if diff >= 0 {
            allNegative = false
        }

        if diff <= 0 {
            allPositive = false
        } 

        abs_diff := max(diff, -diff)
        if abs_diff < 1 {
            minOne = false
        }

        if abs_diff > 3 {
            maxThree = false
        }
    }
    return (allPositive || allNegative) && maxThree && minOne
}

File: src/year2024/day2/scala/part2.scala

@main def part2(): Int  = {
  val lines = scala.io.Source.fromFile("../prod2.txt").getLines().toList
  var ans: Int = 0

  for (line <- lines) {
    var safe: Boolean = false
    val nums = line.split(" +").map(_.toInt).toList
    if (Helpers.is_safe(nums)) {
      safe = true
    } else {
      for (i <- 0 until nums.length) {
        val new_nums = nums.slice(0,i) ++ nums.slice(i+1,nums.length)
        if (Helpers.is_safe(new_nums)) {
          safe = true
        }
      }
    }

    if (safe) {
      ans += 1
    }
  }

  ans
}


object Helpers {
  def is_safe(nums: List[Int]): Boolean = {
    var allPositive = true
    var allNegative = true
    var maxThree = true
    var minOne = true

    for (i <- 1 until nums.length) {
      val diff = nums(i-1) - nums(i)

      if (diff <= 0) {
        allPositive = false
      }

      if (diff >= 0) {
        allNegative = false
      }

      val abs_diff = diff.max(-diff)

      if (abs_diff < 1) {
        minOne = false

      }

      if (abs_diff > 3) {
        maxThree = false
      }
    }

    (allNegative || allPositive) && maxThree && minOne
  }
}
